### DISCUSSION TITLE
Configuring Rubocop

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#6504] [Show and tell]
If you are looking to lint your Ruby code with Rubocop, you can configure it using one of the following settings snippets:

## with bundler

```
    "language_overrides": {
        "Ruby": {
            "format_on_save": "on",
            "formatter": {
                "external": {
                    "command": "./bin/bundle",
                    "arguments": [
                        "exec",
                        "rubocop",
                        "--auto-correct",
                        "--stdin",
                        "corrected.rb",
                        "--stderr"
                    ]
                }
            }
        }
    }
```

## without bundler

```
    "language_overrides": {
        "Ruby": {
            "format_on_save": "on",
            "formatter": {
                "external": {
                    "command": "rubocop",
                    "arguments": [
                        "--auto-correct",
                        "--stdin",
                        "corrected.rb",
                        "--stderr"
                    ]
                }
            }
        }
    }
```

Tip originally posted [here](https://github.com/zed-industries/community/issues/113#issuecomment-1446733321) by [Kucho](https://github.com/kucho).

**Comment by ufchrisg** [comment]
This doesn't appear to work for me. I've tried both the bundle and non bundle version. I can run it in command line (without the --stdin). Does Zed have a console to see if there's any errors in executing the command?

**Reply by JosephTLyons** [reply]
There is an error log if you open the command palette and run `zed: open log`. This only shows the last 1000 lines, so if you need more, you need to check the file itself at `~/Library/Logs/Zed/Zed.log`.  That being said, I'm not sure if the error will surface there or not.

**Reply by ufchrisg** [reply]
Awesome, thank you and it did surface there.

**Reply by ufchrisg** [reply]
Coming back to this. The formatter when called doesn't seem to honor the project directory as I see 
```
2023-03-13T15:08:50 [ERROR] failed to format via external command "./bin/bundle"

Caused by:
    No such file or directory (os error 2)
```

I launched Zed using `zed .` from the project directory but I'm not sure where it's running this command from. I'd hate to have to put my project directory in the path for this to work.

**Reply by ufchrisg** [reply]
Just using `bundle` instead of `./bin/bundle` it finds the bundler that it's in the path but I get the expected error of not finding the gemfile since it's definitely not running from the project dir
```
2023-03-13T15:12:22 [ERROR] failed to format via external command "bundle"

Caused by:
    command failed with exit code Some(10):
    stdout: 
    stderr: Could not locate Gemfile or .bundle/ directory
```

**Reply by ufchrisg** [reply]
This resolved with an update from 0.75.2 to 0.77.3. Didn't see anything regarding this in the notes but I'll take it.

**Comment by btihen** [comment]
the log trick helped - thanks, but how do I get rubocop to show/suggest ideas in the editor - without looking in the logs - like .map { |a| a.empty? } as `.map(&:empty?)`

**Reply by tbsvttr** [reply]
@btihen That is a very good question! For me it also just dumps the result of running to Rubocop command in terminal into the log. But obviously I want to have the autocorrect applied to the file and the linting suggestions recommended in file view.

**Comment by bobmaerten** [comment]
I had trouble making this work too, so I share here as it might help someone too.
Runnin rubocop autocorrect with errors/warning in the requested file will return a nonzero exit code. Which will make formatting fail.
A trick that would prevent this, is to use the `--fail-level` argument:
```
"language_overrides": {
        "Ruby": {
            "format_on_save": "on",
            "formatter": {
                "external": {
                    "command": "bundle",
                    "arguments": [
                        "exec",
                        "rubocop",
                        "--fail-level",
                        "error",
                        "--autocorrect",
                        "--stdin",
                        "corrected.rb",
                        "--stderr"
                    ]
                }
            }
        }
    }
```

**Comment by janko** [comment]
Adding the `--server` argument will also significantly speed up formatting, as it will start a Rubocop server in the background 👌🏻

**Comment by jmonteiro** [comment]
As a small update and reflecting previous comments and suggestions, on Zed 0.14.0.5 I've been successfully using this in `settings.json` for projects bundled with ruby-lsp and rubocop:

```json
{
  "languages": {
    "Ruby": {
      "tab_size": 2,
      "hard_tabs": false,
      "format_on_save": "on",
      "language_servers": ["ruby-lsp", "!solargraph"],
      "formatter": {
        "external": {
          "command": "bundle",
          "arguments": [
            "exec",
            "rubocop",
            "--server",
            "--fail-level",
            "error",
            "--autocorrect",
            "--stdin",
            "corrected.rb",
            "--stderr"
          ]
        }
      }
    }
  }
}
```

**Comment by benjaminwelch** [comment]
I'm still running into this issue. I've tried changing the command but none of the suggestions have worked for me.

Here's what I have in my `settings.json` file:
```json
 "languages": {
    "Ruby": {
      "language_servers": [
        "tailwindcss-language-server",
        "ruby-lsp",
        "!solargraph"
      ],
      "format_on_save": "on",
      "formatter": {
        "external": {
          "command": "bundle",
          "arguments": [
            "exec",
            "rubocop",
            "--server",
            "--fail-level",
            "error",
            "--autocorrect",
            "--stdin",
            "corrected.rb",
            "--stderr"
          ]
        }
      }
    },
```
I might be missing something obvious. Really appreciate any suggestions! And I love Zed.

**Reply by ehutzelman** [reply]
I had to move the formatter block into "language_overrides" instead of "languages" for it to work. As shown in https://github.com/zed-industries/zed/discussions/6504#discussioncomment-5366593

**Reply by orbulant** [reply]
Putting it in `language_overrides` somehow magically worked for me too...

**Comment by orbulant** [comment]
I had notice that if you have ruby-lsp, not following the official docs worked for me for some reason probably due to how the `ruby-lsp` has rubocop attached in some way? 

Before
```js
{
  "languages": {
    "Ruby": {
      "language_servers": ["ruby-lsp", "rubocop", "!solargraph", "..."]
    }
  },
  "lsp": {
    "rubocop": {
      "initialization_options": {
        "safeAutocorrect": false
      }
    },
    "ruby-lsp": {
      "initialization_options": {
        "enabledFeatures": {
          "diagnostics": false
        }
      }
    }
  }
}
```

After

```js
{
  "languages": {
    "Ruby": {
      "language_servers": ["ruby-lsp", "!solargraph", "..."] // Removed `rubocop`
    }
  },
  // Removed `lsp` block
}
```

This somehow worked for me coming from a VSCode, rbenv setup
